src/network/network_coordinator.cpp

```diff
@@ -0,0 +1,701 @@
+
+	void OnFailure() override
+	{
+		_network_coordinator_client.connecting = false;
+		_network_coordinator_client.CloseConnection(true);
+	}
+
+	void OnConnect(SOCKET s) override
+	{
+		assert(_network_coordinator_client.sock == INVALID_SOCKET);
+
+		_network_coordinator_client.sock = s;
+		_network_coordinator_client.last_activity = std::chrono::steady_clock::now();
+		_network_coordinator_client.connecting = false;
+	}
+};
+
+bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_ERROR(Packet *p)
+{
+	NetworkCoordinatorErrorType error = (NetworkCoordinatorErrorType)p->Recv_uint8();
+	std::string detail = p->Recv_string(NETWORK_ERROR_DETAIL_LENGTH);
+
+	switch (error) {
+		case NETWORK_COORDINATOR_ERROR_UNKNOWN:
+			this->CloseConnection();
+			return false;
+
+		case NETWORK_COORDINATOR_ERROR_REGISTRATION_FAILED:
+			SetDParamStr(0, detail);
+			ShowErrorMessage(STR_NETWORK_ERROR_COORDINATOR_REGISTRATION_FAILED, STR_JUST_RAW_STRING, WL_ERROR);
+
+			/* To prevent that we constantly try to reconnect, switch to local game. */
+			_settings_client.network.server_game_type = SERVER_GAME_TYPE_LOCAL;
+
+			this->CloseConnection();
+			return false;
+
+		case NETWORK_COORDINATOR_ERROR_INVALID_JOIN_KEY: {
+			std::string join_key = "+" + detail;
+
+			/* Find the connecter based on the join-key. */
+			auto connecter_it = this->connecter_pre.find(join_key);
+			if (connecter_it == this->connecter_pre.end()) return true;
+			this->connecter_pre.erase(connecter_it);
+
+			/* Mark the server as offline. */
+			NetworkGameList *item = NetworkGameListAddItem(join_key);
+			item->online = false;
+
+			UpdateNetworkGameWindow();
+			return true;
+		}
+
+		default:
+			Debug(net, 0, "Invalid error type {} received from Game Coordinator", error);
+			this->CloseConnection();
+			return false;
+	}
+}
+
+bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_REGISTER_ACK(Packet *p)
+{
+	_settings_client.network.server_join_key = "+" + p->Recv_string(NETWORK_JOIN_KEY_LENGTH);
+	_settings_client.network.server_join_key_secret = p->Recv_string(NETWORK_JOIN_KEY_SECRET_LENGTH);
+	_network_server_connection_type = (ConnectionType)p->Recv_uint8();
+
+	if (_network_server_connection_type == CONNECTION_TYPE_ISOLATED) {
+		ShowErrorMessage(STR_NETWORK_ERROR_COORDINATOR_ISOLATED, STR_NETWORK_ERROR_COORDINATOR_ISOLATED_DETAIL, WL_ERROR);
+	}
+
+	/* Users can change the join-key in the settings, but this has no effect
+	 * on the join-key as assigned by the server. So _network_game_info.join_key
+	 * contains the current join-key, and _settings_client.network.server_join_key
+	 * contains the one we will attempt to re-use when registering again. */
+	_network_game_info.join_key = _settings_client.network.server_join_key;
+
+	SetWindowDirty(WC_CLIENT_LIST, 0);
+
+	if (_network_dedicated) {
+		std::string connection_type;
+		switch (_network_server_connection_type) {
+			case CONNECTION_TYPE_ISOLATED: connection_type = "!! Remote connections blocked !!"; break;
```

Comment:
We warn a server admin here that the remote connections are blocked. I don't think this fits together well with the message that comes just after. I think this should be a more noticeable warning message that players cannot even join the server, even though the server admin does indeed receive the message that their server is registered with the coordinator.

Reply:
Suggestions?

Reply:
First thing that comes to mind is a bold warning in yellow text.

Reply:
Bit the issue I had with this: normally, even for the worst connections, it falls back to TURN. So this means something is going terrible wrong. And I have a hard time getting this message across without being all technical and stuff. So I was looking for something simple and to the point, which will cause people to look for what is going on on the forums etc ..

but that message is being a pita for me, as I can't find wording that works :) So any input on the text itself would be much appreciated (not only looking at you, but at anyone :P)

Reply:
Might I suggest `None (all incoming connections blocked)` as the message? I think that gets the point across even to non-technical players.

Reply:
That sounds a lot better! Tnx :D

The only improvement I would fancy is if we can make it clear we are not the ones doing the blocking .. but without making a very lengthy sentence, that is tricky ;)

Reply:
`None (blocked by firewall/network configuration)`?